beekeeper-scheduler/src/main/java/com/expediagroup/beekeeper/scheduler/service/ExpiredHousekeepingMetadataSchedulerService.java

```diff
@@ -79,6 +83,61 @@ private HousekeepingMetadata createOrUpdateHousekeepingMetadata(HousekeepingMeta
+      compareMaxCleanupTimestampWithTable(existingHousekeepingMetadata);
+    }
+
     return existingHousekeepingMetadata;
   }
+
+  /**
+   * When the cleanup delay of a table with partitions is altered, the delay should be changed but the cleanup timestamp
+   * should be the max timestamp of any of the partitions which the table has.
+   *
+   * i.e. if the cleanup delay was 10 but now its being updated to 2, the delay should match any partition which has
+   * delay 10 (or above) to prevent premature attempts to cleanup the table.
+   *
+   * @param housekeepingMetadata
+   */
+  private void compareMaxCleanupTimestampWithTable(HousekeepingMetadata housekeepingMetadata) {
+    LocalDateTime currentCleanupTimestamp = housekeepingMetadata.getCleanupTimestamp();
+    LocalDateTime maxCleanupTimestamp = housekeepingMetadataRepository.findMaximumCleanupTimestampForDbAndTable(
+        housekeepingMetadata.getDatabaseName(), housekeepingMetadata.getTableName());
+
+    if (maxCleanupTimestamp != null && maxCleanupTimestamp.isAfter(currentCleanupTimestamp)) {
+      log.info("Updating entry for \"{}.{}\". Cleanup timestamp is now \"{}\".", housekeepingMetadata.getDatabaseName(),
+          housekeepingMetadata.getTableName(), maxCleanupTimestamp);
+      housekeepingMetadata.setCleanupTimestamp(maxCleanupTimestamp);
+    }
+  }
+
+  /**
+   * When a new partition is added to a table, check to see if the new cleanup delay will be later than the current
+   * cleanup delay for the table.
+   * The cleanup timestamp of a partitioned table should be equivalent to that of the last partition which will be
+   * dropped to prevent premature attempts to cleanup the table.
+   *
+   * @param partitionMetadata
+   */
+  private void comparePartitionCleanupTimestampWithTable(HousekeepingMetadata partitionMetadata) {
+    HousekeepingMetadata tableMetadata = housekeepingMetadataRepository.findRecordForCleanupByDbTableAndPartitionName(
+        partitionMetadata.getDatabaseName(), partitionMetadata.getTableName(), null).get();
+
+    LocalDateTime partitionCleanupTimestamp = partitionMetadata.getCleanupTimestamp();
+    LocalDateTime currentCleanupTimestamp = tableMetadata.getCleanupTimestamp();
+
+    if (partitionCleanupTimestamp.isAfter(currentCleanupTimestamp)) {
+      log.info("Updating entry for \"{}.{}\". Cleanup timestamp is now \"{}\".", tableMetadata.getDatabaseName(),
+          tableMetadata.getTableName(), partitionCleanupTimestamp);
+      tableMetadata.setCleanupTimestamp(partitionCleanupTimestamp);
+      housekeepingMetadataRepository.save(tableMetadata);
+    }
+  }
+
+  private boolean isPartitionedTable(HousekeepingMetadata housekeepingMetadata) {
+    Long numPartitions = housekeepingMetadataRepository.countRecordsForGivenDatabaseAndTableWherePartitionIsNotNull(
+        housekeepingMetadata.getDatabaseName(), housekeepingMetadata.getTableName());
+
+    return numPartitions > 0 && housekeepingMetadata.getPartitionName() == null;
```

Comment:
Just curious, why do we also need to check that the partition name is null?

Reply:
This method is called when we're updating an existing record and we only want to update a record for a table. So numPartitions tells us if there are any existing partitions on that table, and the null check makes sure the housekeepingMetadata record being updated is for the table and not one of its partitions.